### ISSUE TITLE
Should merging ambient external modules share a scope that includes their export specifiers?

### ISSUE THREAD

**Issue by JsonFreeman** [Microsoft/TypeScript#3535]
When namespaces merge, any members exported by one namespace declaration are in scope in all the merged declarations. Ambient external modules like `declare module "foo" { }` can merge too. It seems natural then that exported declarations are in scope in merged external modules too. But external modules have additional ways of exporting their members, via export specifiers, export default, and `export *`. Should export specifiers contribute to the shared export scope as well?

``` ts
declare module 'm' {
     export interface I { }
     interface J { }
     interface K { }
     export { K };
     interface L { }
     export { L as L2 };
     export { P } from 'other';
}
declare module 'm' {
     export function f(): I; // in scope because of export modifier
     export function g(): J; // not in scope (lacks export modifier)
     export function h(): K; // should this be in scope?
     export function h(): L; // should this be in scope?
     export function h(): P; // should this be in scope?
}
declare module 'other' {
     export interface P { }
}
```

Today only J is in scope.

The key difference between export modifiers and export specifiers is that when an export modifier is used, the exporting entity is guaranteed to have a corresponding local entity with the same name. This is not true of export specifiers, as they can export stuff that is not declared locally, or they can export using a different name.

I can think of 4 options. I've listed them in order from smallest export scope to largest export scope:
1. Shared export scopes are not created at all for external modules. This is a pretty harsh breaking change.
2. Only an export modifier (implicit or explicit) can put the name into a shared export scope, but an export specifier cannot. This is the current behavior. It breaks a nice user intuition that using the export modifier is equivalent to leaving it out and writing an export specifier.
3. An export specifier that exports a locally declared module member with the same name is also in scope and treated just like an export modifier. This sounds like an oddly specific rule. But it maintains the user intuition that option 2 breaks, and when you think about it, it seems pretty natural I think. This would mean that in the above example, I and K are in scope.
4. All exported entities in a module are in the shared export scope, no matter how they were exported. This is inconsistent with ES6 export semantics.

I think 1 and 4 are not viable options. 3 seems ideal, but I could also see a case for 2. How about it?

**Comment by benliddicott** [comment]
It seems to me that the names `P`, `K` and `L2` ought to be in scope in the second `declare module 'm'`, but `L` should not. So it should be possible to write:

    export function h(): L2; // should this be in scope? Yes.


I suppose it is not a big deal provided you can work around it by re-importing and re-exporting them - but would that produce an error due to duplicate entries?

**Comment by JsonFreeman** [comment]
I disagree that P should be in scope. I think that everything in scope in the second declaration should also be in scope in the first, and P is not in scope in the first.

K makes sense to put in scope, and I can kind of see L being in scope in the first declaration, but L2 being in scope in the second declaration in place of L.

Depending on how you do the re-imports and re-exports, you may or may not get an error, but I think for the most part you should be safe.

**Comment by benliddicott** [comment]
Just played around with it a bit. Sorry to be so stupid, but since they all work if you reimport the symbol, how much does it matter?

    import {K} from 'm';
    export function h(): K; // In scope now
    import {L2 as L} from 'm';
    export function h(): L; // In scope now
    import {P} from 'm';
    export function h(): P; // In scope now

If they became impossible to use then it would be an issue, bit they don't. 

There is a nice symmetry between export specifier and the import specifier. For the user's intuition, they can think of it as if an export *modifier* creates an implicit import whenever the module is reopened, but an export specifier does not. This makes sense because an export specifier can create an alias but a modifier cannot. That's how I'm going to remember it anyway.

In other words I am coming round to the view that 2) current situation is best. It's less complicated (and presumably less work since it is already done) and needn't unduly shock the user's intuitions.

**Comment by JsonFreeman** [comment]
Okay @benliddicott, that is fair. Thanks for your opinion and for your suggested explanation of the behavior.

**Comment by RyanCavanaugh** [comment]
Seems like the answer is "no"
